File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/FrequencyValueBase.vb
'*******************************************************************************************************
'  FrequencyValueBase.vb - Frequency and DfDt value base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  11/12/2004 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization

''' <summary>This class represents the protocol independent a frequency and dfdt value.</summary>
<CLSCompliant(False), Serializable()> _
Public MustInherit Class FrequencyValueBase

    Inherits ChannelValueBase(Of IFrequencyDefinition)
    Implements IFrequencyValue

    Private m_frequency As Single
    Private m_dfdt As Single

    <Serializable()> _
    Public Enum CompositeValueType
        Frequency
        DfDt
    End Enum

    Protected Sub New()
    End Sub

    Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

        MyBase.New(info, context)

        ' Deserialize frequency value
        m_frequency = info.GetSingle("frequency")
        m_dfdt = info.GetSingle("dfdt")

    End Sub

    Protected Sub New(ByVal parent As IDataCell)

        MyBase.New(parent)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal frequency As Single, ByVal dfdt As Single)
    Protected Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal frequency As Single, ByVal dfdt As Single)

        MyBase.New(parent, frequencyDefinition)

        m_frequency = frequency
        m_dfdt = dfdt

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal unscaledFrequency As Int16, ByVal unscaledDfDt As Int16)
    Protected Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal unscaledFrequency As Int16, ByVal unscaledDfDt As Int16)

        MyClass.New(parent, frequencyDefinition, _
            Convert.ToSingle(unscaledFrequency / frequencyDefinition.ScalingFactor + frequencyDefinition.Offset), _
            Convert.ToSingle(unscaledDfDt / frequencyDefinition.DfDtScalingFactor + frequencyDefinition.DfDtOffset))

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)
    Protected Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(parent, frequencyDefinition)
        ParseBinaryImage(Nothing, binaryImage, startIndex)

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal frequencyValue As IFrequencyValue)
    Protected Sub New(ByVal frequencyValue As IFrequencyValue)

        MyClass.New(frequencyValue.Parent, frequencyValue.Definition, frequencyValue.Frequency, frequencyValue.DfDt)

    End Sub

    Public Overridable Property Frequency() As Single Implements IFrequencyValue.Frequency
        Get
            Return m_frequency
        End Get
        Set(ByVal value As Single)
            m_frequency = value
        End Set
    End Property

    Public Overridable Property DfDt() As Single Implements IFrequencyValue.DfDt
        Get
            Return m_dfdt
        End Get
        Set(ByVal value As Single)
            m_dfdt = value
        End Set
    End Property

    Public Overridable Property UnscaledFrequency() As Int16 Implements IFrequencyValue.UnscaledFrequency
        Get
            With Definition
                Return Convert.ToInt16((m_frequency - .Offset) * .ScalingFactor)
            End With
        End Get
        Set(ByVal value As Int16)
            With Definition
                m_frequency = value / .ScalingFactor + .Offset
            End With
        End Set
    End Property

    Public Overridable Property UnscaledDfDt() As Int16 Implements IFrequencyValue.UnscaledDfDt
        Get
            With Definition
                Return Convert.ToInt16((m_dfdt - .DfDtOffset) * .DfDtScalingFactor)
            End With
        End Get
        Set(ByVal value As Int16)
            With Definition
                m_dfdt = value / .DfDtScalingFactor + .DfDtOffset
            End With
        End Set
    End Property

    Default Public Overrides Property CompositeValue(ByVal index As Integer) As Single
        Get
            Select Case index
                Case CompositeValueType.Frequency
                    Return m_frequency
                Case CompositeValueType.DfDt
                    Return m_dfdt
                Case Else
                    Throw New IndexOutOfRangeException("Specified frequency value composite index, " & index & ", is out of range - there are only two composite values for a frequency value: frequency (0) and df/dt (1)")
            End Select
        End Get
        Set(ByVal value As Single)
            Select Case index
                Case CompositeValueType.Frequency
                    m_frequency = value
                Case CompositeValueType.DfDt
                    m_dfdt = value
                Case Else
                    Throw New IndexOutOfRangeException("Specified frequency value composite index, " & index & ", is out of range - there are only two composite values for a frequency value: frequency (0) and df/dt (1)")
            End Select
        End Set
    End Property

    Public Overrides ReadOnly Property CompositeValueCount() As Integer
        Get
            Return 2
        End Get
    End Property

    Public Overrides ReadOnly Property IsEmpty() As Boolean
        Get
            Return (m_frequency = 0 And m_dfdt = 0)
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyLength() As UInt16
        Get
            If DataFormat = Phasors.DataFormat.FixedInteger Then
                Return 4
            Else
                Return 8
            End If
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyImage() As Byte()
        Get
            Dim buffer As Byte() = CreateArray(Of Byte)(BodyLength)

            If DataFormat = Phasors.DataFormat.FixedInteger Then
                EndianOrder.BigEndian.CopyBytes(UnscaledFrequency, buffer, 0)
                EndianOrder.BigEndian.CopyBytes(UnscaledDfDt, buffer, 2)
            Else
                EndianOrder.BigEndian.CopyBytes(m_frequency, buffer, 0)
                EndianOrder.BigEndian.CopyBytes(m_dfdt, buffer, 4)
            End If

            Return buffer
        End Get
    End Property

    Protected Overrides Sub ParseBodyImage(ByVal state As IChannelParsingState, ByVal binaryImage() As Byte, ByVal startIndex As Integer)

        If DataFormat = Phasors.DataFormat.FixedInteger Then
            UnscaledFrequency = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex)
            UnscaledDfDt = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 2)
        Else
            m_frequency = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex)
            m_dfdt = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex + 4)
        End If

    End Sub

    Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

        MyBase.GetObjectData(info, context)

        ' Serialize frequency value
        info.AddValue("frequency", m_frequency)
        info.AddValue("dfdt", m_dfdt)

    End Sub

    Public Overrides ReadOnly Property Attributes() As System.Collections.Generic.Dictionary(Of String, String)
        Get
            With MyBase.Attributes
                .Add("Frequency Value", Frequency)
                .Add("df/dt Value", DfDt)
                .Add("Unscaled Frequency Value", UnscaledFrequency)
                .Add("Unscaled df/dt Value", UnscaledDfDt)
            End With

            Return MyBase.Attributes
        End Get
    End Property

End Class

File: Source/Libraries/GSF.PhasorProtocols/FrequencyValueBase.cs
//******************************************************************************************************
//  FrequencyValueBase.cs - Gbtc
//
//  Copyright © 2012, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://www.opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  11/12/2004 - J. Ritchie Carroll
//       Generated original version of source code.
//  08/07/2009 - Josh L. Patterson
//       Edited Comments.
//  09/15/2009 - Stephen C. Wills
//       Added new header and license agreement.
//  12/17/2012 - Starlynn Danyelle Gilliam
//       Modified Header.
//
//******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;

namespace GSF.PhasorProtocols
{
    #region [ Enumerations ]

    /// <summary>
    /// Composite frequency value indices enumeration.
    /// </summary>
    public enum CompositeFrequencyValue
    {
        /// <summary>
        /// Composite frequency value index.
        /// </summary>
        Frequency,
        /// <summary>
        /// Composite dF/dt value index.
        /// </summary>
        DfDt
    }

    #endregion

    /// <summary>
    /// Represents the common implementation of the protocol independent representation of a frequency and dF/dt value.
    /// </summary>
    [Serializable]
    public abstract class FrequencyValueBase : ChannelValueBase<IFrequencyDefinition>, IFrequencyValue
    {
        #region [ Members ]

        // Fields
        private double m_frequency;
        private double m_dfdt;
        private bool m_frequencyAssigned;
        private bool m_dfdtAssigned;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="FrequencyValueBase"/>.
        /// </summary>
        /// <param name="parent">The <see cref="IDataCell"/> parent of this <see cref="FrequencyValueBase"/>.</param>
        /// <param name="frequencyDefinition">The <see cref="IFrequencyDefinition"/> associated with this <see cref="FrequencyValueBase"/>.</param>
        protected FrequencyValueBase(IDataCell parent, IFrequencyDefinition frequencyDefinition)
            : base(parent, frequencyDefinition)
        {
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyValueBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IDataCell"/> parent of this <see cref="FrequencyValueBase"/>.</param>
        /// <param name="frequencyDefinition">The <see cref="IFrequencyDefinition"/> associated with this <see cref="FrequencyValueBase"/>.</param>
        /// <param name="frequency">The floating point value that represents this <see cref="FrequencyValueBase"/>.</param>
        /// <param name="dfdt">The floating point value that represents the change in this <see cref="FrequencyValueBase"/> over time.</param>
        protected FrequencyValueBase(IDataCell parent, IFrequencyDefinition frequencyDefinition, double frequency, double dfdt)
            : base(parent, frequencyDefinition)
        {
            m_frequency = frequency;
            m_dfdt = dfdt;

            m_frequencyAssigned = !double.IsNaN(frequency);
            m_dfdtAssigned = !double.IsNaN(dfdt);
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyValueBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected FrequencyValueBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize frequency value
            m_frequency = info.GetDouble("frequency");
            m_dfdt = info.GetDouble("dfdt");

            m_frequencyAssigned = true;
            m_dfdtAssigned = true;
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets or sets the floating point value that represents this <see cref="FrequencyValueBase"/>.
        /// </summary>
        public virtual double Frequency
        {
            get => m_frequency;
            set
            {
                m_frequency = value;
                m_frequencyAssigned = true;
            }
        }

        /// <summary>
        /// Gets or sets the floating point value that represents the change in this <see cref="FrequencyValueBase"/> over time.
        /// </summary>
        public virtual double DfDt
        {
            get => m_dfdt;
            set
            {
                m_dfdt = value;
                m_dfdtAssigned = true;
            }
        }

        /// <summary>
        /// Gets or sets the unscaled integer representation of this <see cref="FrequencyValueBase"/>.
        /// </summary>
        public virtual int UnscaledFrequency
        {
            get
            {
                unchecked
                {
                    return (int)((m_frequency - Definition.Offset) * Definition.ScalingValue);
                }
            }
            set
            {
                m_frequency = value / (double)Definition.ScalingValue + Definition.Offset;
                m_frequencyAssigned = true;
            }
        }

        /// <summary>
        /// Gets or sets the unscaled integer representation of the change in this <see cref="FrequencyValueBase"/> over time.
        /// </summary>
        public virtual int UnscaledDfDt
        {
            get
            {
                unchecked
                {
                    return (int)((m_dfdt - Definition.DfDtOffset) * Definition.DfDtScalingValue);
                }
            }
            set
            {
                m_dfdt = value / (double)Definition.DfDtScalingValue + Definition.DfDtOffset;
                m_dfdtAssigned = true;
            }
        }

        /// <summary>
        /// Gets boolean value that determines if none of the composite values of <see cref="FrequencyValueBase"/> have been assigned a value.
        /// </summary>
        /// <returns>True, if no composite values have been assigned a value; otherwise, false.</returns>
        public override bool IsEmpty => !m_frequencyAssigned || !m_dfdtAssigned;

        /// <summary>
        /// Gets total number of composite values that this <see cref="FrequencyValueBase"/> provides.
        /// </summary>
        public override int CompositeValueCount => 2;

        /// <summary>
        /// Gets the length of the <see cref="BodyImage"/>.
        /// </summary>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override int BodyLength
        {
            get
            {
                if (DataFormat == DataFormat.FixedInteger)
                    return 4;
                return 8;
            }
        }

        /// <summary>
        /// Gets the binary body image of the <see cref="FrequencyValueBase"/> object.
        /// </summary>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override byte[] BodyImage
        {
            get
            {
                byte[] buffer = new byte[BodyLength];

                // Had to make a decision on usage versus typical protocol implementation when
                // exposing values as double / int when protocols typically use float / short for
                // transmission. Exposing values as double / int makes class more versatile by
                // allowing future protocol implementations to support higher resolution values
                // simply by overriding BodyLength, BodyImage and ParseBodyImage. However, exposing
                // the double / int values runs the risk of providing values that are outside the
                // data type limitations, hence the unchecked section below. Risk should generally
                // be low in typical usage scenarios since values being transmitted via a generated
                // image were likely parsed previously from a binary image with the same constraints.
                unchecked
                {
                    if (DataFormat == DataFormat.FixedInteger)
                    {
                        BigEndian.CopyBytes((short)UnscaledFrequency, buffer, 0);
                        BigEndian.CopyBytes((short)UnscaledDfDt, buffer, 2);
                    }
                    else
                    {
                        BigEndian.CopyBytes((float)m_frequency, buffer, 0);
                        BigEndian.CopyBytes((float)m_dfdt, buffer, 4);
                    }
                }

                return buffer;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="FrequencyValueBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Frequency Value", Frequency.ToString());
                baseAttributes.Add("dF/dt Value", DfDt.ToString());
                baseAttributes.Add("Unscaled Frequency Value", UnscaledFrequency.ToString());
                baseAttributes.Add("Unscaled dF/dt Value", UnscaledDfDt.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Gets the specified composite value of this <see cref="FrequencyValueBase"/>.
        /// </summary>
        /// <param name="index">Index of composite value to retrieve.</param>
        /// <remarks>
        /// Some <see cref="ChannelValueBase{T}"/> implementations can contain more than one value, this method is used to abstractly expose each value.
        /// </remarks>
        /// <returns>A <see cref="double"/> representing the composite value.</returns>
        public override double GetCompositeValue(int index)
        {
            switch (index)
            {
                case (int)CompositeFrequencyValue.Frequency:
                    return m_frequency;
                case (int)CompositeFrequencyValue.DfDt:
                    return m_dfdt;
                default:
                    throw new ArgumentOutOfRangeException(nameof(index), "Invalid composite index requested");
            }
        }

        /// <summary>
        /// Parses the binary body image.
        /// </summary>
        /// <param name="buffer">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="buffer"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="buffer"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override int ParseBodyImage(byte[] buffer, int startIndex, int length)
        {
            // Length is validated at a frame level well in advance so that low level parsing routines do not have
            // to re-validate that enough length is available to parse needed information as an optimization...

            if (DataFormat == DataFormat.FixedInteger)
            {
                UnscaledFrequency = BigEndian.ToInt16(buffer, startIndex);
                UnscaledDfDt = BigEndian.ToInt16(buffer, startIndex + 2);

                return 4;
            }

            m_frequency = BigEndian.ToSingle(buffer, startIndex);
            m_dfdt = BigEndian.ToSingle(buffer, startIndex + 4);

            m_frequencyAssigned = true;
            m_dfdtAssigned = true;

            return 8;
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize frequency value
            info.AddValue("frequency", m_frequency);
            info.AddValue("dfdt", m_dfdt);
        }

        #endregion
    }
}